sycl/doc/extensions/reduction/reduction.md

```diff
@@ -0,0 +1,147 @@
+
+To enable compile-time specialization of reduction algorithms, an implementation may define additional template arguments to the `reducer` class.  The `reducer` type for a given reduction can be inspected using `decltype(reduction(var, identity, combiner))::reducer_type`.
+
+# Adding `reduction` Objects to `parallel_for`
+
+```c++
+template <typename KernelName, typename KernelType, int dimensions, typename... Rest>
+void parallel_for(range<dimensions>numWorkItems, Rest&&... rest);
+```
+
+The `rest` parameter pack consists of 0 or more `reduction` objects followed by the kernel functor.  For each `reduction` object operating on values of type `T`, the kernel functor should take an additional parameter of type `reducer<T, BinaryOperation>&`.  For convenience and to avoid supplying the same information twice, it is expected that developers using C++14 will typically make use of `auto&` in place of specifying the reducer type.
+
+The implementation must guarantee that it is safe for each concurrently executing work-item to call the `combine` function of a reducer in parallel.  An implementation is free to re-use reducer variables (e.g. across work-groups scheduled to the same compute unit) if it can guarantee that it is safe to do so.
+
+The combination order of different reducers is unspecified, as are when and how the value of each reducer is combined with the original variable.  The value of the original variable at any point during execution of the kernel is undefined, and the final value is only visible after the kernel completes.
+
+## Example
+```c++
+// Compute a dot-product by reducing all computed values using standard plus functor
+queue.submit([&](handler& cgh)
+{
+    auto a = a_buf.get_access<access::mode::read>(cgh);
+    auto b = b_buf.get_access<access::mode::read>(cgh);
+    auto sum = sum_buf.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class dot_product>(nd_range<1>{N, M}, reduction(sum, 0, plus<int>()), [=](nd_item<1> it, auto& sum)
+    {
+        int i = it.get_global_id(0);
+        sum += (a[i] * b[i]);
+    });
+});
+```
+
+# Array Reductions
+
+SYCL buffers do not distinguish between scalars and arrays in the type system; a scalar in device memory must be represented as a buffer of size 1.  This proposal assumes that the majority of reductions are scalar, and that a buffer passed to a reduction should therefore always be interpreted as a reduction of a single element.  In order to support reductions of array types, we propose to treat them as a special-case: the user must explicitly request an array reduction by passing a `span` denoting the region of the buffer to include in the reduction.  The semantics of an array reduction of size N should be equivalent to N independent reductions.
```

Comment:
You could make a proposal pushing 0-dimension buffers for storing scalars...

Reply:
Thanks @keryell, this is a good suggestion.  I'll start work on a separate proposal for 0-dimension buffers, since I agree they would be useful.

Rather than make this proposal dependent on that one, we could say that a 0-dimension accessor (which already exists) creates a scalar reduction and a non-0-dimension accessor creates an array reduction.  0-dimension buffers wouldn't be required, but would simplify the creation of 0-dimension accessors.

We'd still need to support the `span` version of things for compatibility with [USM pointers](https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/usm), but that requirement could be made explicit.

What do you think?  If this sounds okay to you I'll update the proposal.

Reply:
Using `span` seems more secure. And it is looks like C++20. :-)
But in the meantime, probably using a 0-dimension accessor for scalar seems good.